### PR TITLE
Better CLI error messages on user actions at wrong times and circumstances

### LINKED ISSUE #108: Better error messages for invalid user actions in demo
For example:

* CLI action taken on wrong time
  - If a bidder tries to submit a bid when bidding hasn't started: "Cannot submit bid: bidding has not started yet. Bidding should start in 19s."
  - If a bidder tries to submit a bid when bidding has ended: "Cannot submit bid: bidding has ended 33s ago."
  - If a bidder tries to buy the auction lot: "Cannot buy auction lot yet: bidding will end in 2m14s."
  - If the seller tries to reclaim the auction lot: "Cannot reclaim auction lot yet: voucher will expire in 52s."
* A bidder tries to buy the auction lot but he isn't the winning bidder: "Cannot buy auction lot: the winning bidder is someone else."

### PR DESCRIPTION
**Pull request by uhbif19**
Rebased version of #136 without slashed change on CLI notification on state change

Closes #108 

Some review comments not fixed! Some of them was fixed in @nazrhom PR, if I remember correctly.

### CODE REVIEW

**Comment by MangoIV** [comment]
@uhbif19 this should work now. I have made a couple more simple adjustments, the flake should probably be cleaned up at some point though.

**Comment by MangoIV** [comment]
Seems like I have to fix `linting` and `formatting`

**Comment by MangoIV** [comment]
I updated `fourmolu`, if needed I can revert this; I removed the hacks and fixed hlint, too.

**Comment by MangoIV** [comment]
oups, I was not supposed to merge master :#

File: src/HydraAuction/Tx/Common.hs [hunk 1]
```diff
@@ -293,10 +293,10 @@ scriptUtxos script terms = do
...
 data AutoCreateParams = AutoCreateParams
   { authoredUtxos :: [(SigningKey PaymentKey, UTxO)]
-  , -- | Utxo which TxIns will be used as reference inputs
-    referenceUtxo :: UTxO
-  , -- | Nothing means collateral will be chosen automatically from given UTxOs
-    collateral :: Maybe TxIn
+  , referenceUtxo :: UTxO
+  -- ^ Utxo which TxIns will be used as reference inputs
+  , collateral :: Maybe TxIn
+  -- ^ Nothing means collateral will be chosen automatically from given UTxOs
```
Comment (uhbif19) [inline src/HydraAuction/Tx/Common.hs:299]: This is something fourmolu requires?

**Review by uhbif19 — COMMENTED** [review]
@MangoIV 

Please consider making refactoring/infrastructure fixing in separate PR from feature PR next time.

I am bad at Nix so not sure if I understand its changes correctly. As I understand correctly that now flake checks are used by both Hercules CI and makefile and pre commit hooks?

Thanks for changes on checks stuff.

**Comment by uhbif19** [comment]
@nazrhom 

@MangoIV fixed the time stuff.

What is left in this PR:

* Cleanup stage missing (see old PR review)
* Seconds counting (see old PR review and issue)
* Finishing #165 on top of this PR changes

File: src/HydraAuction/Tx/Common.hs (hunk 1 above)
Comment (MangoIV) [reply src/HydraAuction/Tx/Common.hs:299]: Yeah due to the version upgrade I did.  If you want I can revert it.

**Comment by MangoIV** [comment]
@uhbif19 
> Please consider making refactoring/infrastructure fixing in separate PR from feature PR next time.

Definitely. It's just that the time stuff did require some changes anyway so I did it all at once... :)

File: src/HydraAuction/Tx/Common.hs (hunk 1 above)
Comment (uhbif19) [reply src/HydraAuction/Tx/Common.hs:299]: No, you right, I think we should follow what fourmolu requires.

**Comment by nazrhom** [comment]
@uhbif19 can we merge this one and then I'll rebase the other one on top?

**Comment by uhbif19** [comment]
@nazrhom I think that better to fix point `Cleanup stage missing (see old PR review)` before merging this PR, cuz it missing stage seems like bug to me. Other than that I agree that it is better to merge and rebase on top it.

File: src/HydraAuction/OnChain/StateToken.hs [hunk 2]
```diff
@@ -43,7 +43,7 @@ mkPolicy (EscrowAddress escrowAddressLocal, terms) redeemer ctx =
         && exactlyOneOutputToEscrow
     BurnVoucher ->
       traceIfFalse "Not exactly one Voucher burned" (txInfoMint info == voucherOnlyValue (-1))
-        && checkInterval terms VoucherExpiredStage info
+        && (checkInterval terms VoucherExpiredStage info || checkInterval terms CleanupStage info)
```
Comment (uhbif19) [inline src/HydraAuction/OnChain/StateToken.hs:46]: Nice.

File: src/HydraAuction/Tx/Escrow.hs [hunk 3]
```diff
@@ -372,5 +372,5 @@ sellerReclaims terms = do
             ]
         , toMint = TxMintValueNone
         , changeAddress = sellerAddress
-        , validityBound = (Just $ voucherExpiry terms, Nothing)
+        , validityBound = (Just $ voucherExpiry terms, Just $ cleanup terms)
```
Comment (uhbif19) [inline src/HydraAuction/Tx/Escrow.hs:375]: I think that is wrong, cuz spec says:

> Under the **seller reclaims** redeemer, we enforce that:
> ...
> - The transaction validity interval starts after the voucher expiry time.

**Review by uhbif19 — COMMENTED** [review]
Other than reclaim timing LGTM. Merge conflict due to my merge master into staging, sorry.

File: src/HydraAuction/OnChain/StateToken.hs (hunk 2 above)
Comment (uhbif19) [reply src/HydraAuction/OnChain/StateToken.hs]: I was wrong, this does not work in general case. We need to check that interval belongs to union, not that it belongs to either of intervals.

And by spec here we need only CleanupStage here, that is what Tx code uses for validity, and that is why tests does not fail.

**Pull request merged**